core/src/main/scala/kafka/server/KafkaApis.scala

```diff
@@ -1823,46 +1824,64 @@ class KafkaApis(val requestChannel: RequestChannel,
   def handleDeleteRecordsRequest(request: RequestChannel.Request): Unit = {
     val deleteRecordsRequest = request.body[DeleteRecordsRequest]
 
-    val unauthorizedTopicResponses = mutable.Map[TopicPartition, DeleteRecordsResponse.PartitionResponse]()
-    val nonExistingTopicResponses = mutable.Map[TopicPartition, DeleteRecordsResponse.PartitionResponse]()
+    val unauthorizedTopicResponses = mutable.Map[TopicPartition, DeleteRecordsPartitionResult]()
+    val nonExistingTopicResponses = mutable.Map[TopicPartition, DeleteRecordsPartitionResult]()
     val authorizedForDeleteTopicOffsets = mutable.Map[TopicPartition, Long]()
 
     val authorizedTopics = filterAuthorized(request, DELETE, TOPIC,
-      deleteRecordsRequest.partitionOffsets.asScala.toSeq.map(_._1.topic))
-    for ((topicPartition, offset) <- deleteRecordsRequest.partitionOffsets.asScala) {
+      deleteRecordsRequest.data.topics.asScala.map(_.name))
+    for ((topicPartition, offset) <- deleteRecordsRequest.data.topics.asScala.flatMap(deleteTopic => {
```

Comment:
I wonder if it's more readable having 2 nested for loops. What do you think?

Reply:
To me what makes it less readable is that the expression for the `for` loop is the `flatMap` & `map`. I think the simplest was to improve that would just be to factor out a `val` for that expression. Does that work for you?